p4_symbolic/util/io.cc

```diff
@@ -0,0 +1,59 @@
+// Copyright 2020 Google LLC
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//      http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+// This is a test file for our protobuf specifications of bmv2 json.
+// It reads an input bmv2 json string (usually the output of p4c) via stdin,
+// it parses the string using protobuf, and then dumps the parsed protobuf
+// objects using protobuf text format and json.
+// The dumps are written to output files whose paths are provided as command
+// line arguments.
+
+#include "p4_symbolic/util/io.h"
+
+#include <cerrno>
+#include <fstream>
+#include <streambuf>
+
+#include "absl/strings/str_format.h"
+#include "p4_pdpi/utils/status_utils.h"
+
+namespace p4_symbolic {
+namespace util {
+
+pdpi::StatusOr<std::string> ReadFile(std::string path) {
+  std::ifstream f;
+  f.open(path.c_str());
+  if (f.fail()) {
+    std::string err;
+    switch (errno) {
+      case EACCES:
+      case ENOENT:
+        err = absl::StrFormat("%s: %s", strerror(errno), path);
+        return pdpi::StatusOr<std::string>(
+            absl::Status(absl::StatusCode::kNotFound, err));
+      default:
+        err = absl::StrFormat("Cannot read file %s, errno = %d", path, errno);
+        return pdpi::StatusOr<std::string>(
+            absl::Status(absl::StatusCode::kUnknown, err));
+    }
+  }
+
+  f >> std::noskipws;  // Read whitespaces.
+  std::string output(std::istreambuf_iterator<char>(f),
+                     (std::istreambuf_iterator<char>()));
+  return pdpi::StatusOr<std::string>(output);
```

Comment:
I do not like the implicit conversions at all. Unless there is some explicit google styling rule or something on this, I prefer to keep it explicit. When I looked at the google style guide, they seemed to encourage explicity.

Reply:
I agree with that sentiment in general, but with StatusOr being used so pervasively I think it's reasonable to be a bit implicit for StatusOr only. The thing is that being explicit in all cases can get really out of hand:
```C++
StatusOr<ReallyReallyReallyLongReturnTypeWith<TemplateParameter>>
HypotheticalFunction() {
  // Imagine you have five returns in here.
}
```
But if you agree (and maybe you don't) that being explicit is not reasonable in all cases, then I think it's best to embrace the implicit style and be consistent throughout.

In google3, there are ~1300 occurrences of `return util::StatusOr`. In other words, it's very uncommon. This is admittedly not a strong argument because this project is open source, though.

I think it's ultimately up to you. Curious what @stefanheule thinks?